### PR TITLE
TDL-19587: Resolve ConnectionResetError

### PR DESCRIPTION
**Pull request by hpatel41**
# Description of change
TDL-19587: Resolve ConnectionResetError
- Updated the code to catch errors due to yield and added manual backoff for errors

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_pendo/streams.py [hunk 1]
```diff
@@ -29,6 +29,15 @@
 # timeout request after 300 seconds
 REQUEST_TIMEOUT = 300
 
+FACTOR = 2
+COUNT = 5
```
Comment (NevilParikh14) [inline tap_pendo/streams.py:33]: Can you please remove this variable as it is not used anywhere in the code?

File: tap_pendo/streams.py [hunk 2]
```diff
@@ -29,6 +29,15 @@
...
+
+def to_giveup(e):
```
Comment (dbshah1212) [inline tap_pendo/streams.py:35]: Use error instead of e

File: tap_pendo/streams.py [hunk 3]
```diff
@@ -417,33 +426,44 @@ def lookback_window(self):
...
+                raise Server42xRateLimitError(resp.reason)
+
+            resp.raise_for_status() # Check for requests status and raise exception in failure
+
+            # Get records from raw response
+            for item in ijson.items(resp.raw, 'results.item'):
+                yield humps.decamelize(item)
+            resp.close()
+        except (ConnectionError, ProtocolError, ReadTimeoutError, requests.exceptions.RequestException, Server42xRateLimitError) as e:
+            # Catch requestException and raise error if we have to giveup for certain condiitons
```
Comment (savan-chovatiya) [inline tap_pendo/streams.py:453]:
```suggestion
            # Catch requestException and raise errors if we have to give up for certain conditions
```

File: tap_pendo/streams.py (hunk 1 above)
Comment (dbshah1212) [inline tap_pendo/streams.py:33]: We are defining it but not using

File: tap_pendo/streams.py [hunk 4]
```diff
@@ -417,33 +426,44 @@ def lookback_window(self):
...
+                raise e from None
+            # Raise error if we have retried for 5 times
+            if count == 5:
+                LOGGER.info("Giving up request(...) after 5 tries (%s: %s)", e.__class__.__name__, str(e))
+                raise e from None
+
+            LOGGER.info("Backing off request(...) for %ss (%s: %s)", FACTOR ** count, e.__class__.__name__, str(e))
+            # Sleep for (2 ^ count) seconds ie. 2, 4, 8, 16, 32
+            time.sleep(FACTOR ** count)
+            count += 1
```
Comment (dbshah1212) [inline tap_pendo/streams.py:464]: define count = 5 and decrease it so we need to check `if count` only.

File: tap_pendo/streams.py [hunk 5]
```diff
@@ -29,6 +29,15 @@
 # timeout request after 300 seconds
 REQUEST_TIMEOUT = 300
 
+FACTOR = 2
```
Comment (dbshah1212) [inline tap_pendo/streams.py:32]: Use BACKOFF_FACTOR

File: tap_pendo/streams.py [hunk 6]
```diff
@@ -417,33 +426,44 @@ def lookback_window(self):
...
+            if 'Too Many Requests' in resp.reason:
+                retry_after = 30
+                LOGGER.info("Rate limit reached. Sleeping for %s seconds",
+                            retry_after)
+                time.sleep(retry_after)
+                raise Server42xRateLimitError(resp.reason)
+
+            resp.raise_for_status() # Check for requests status and raise exception in failure
+
+            # Get records from raw response
```
Comment (dbshah1212) [inline tap_pendo/streams.py:448]:
```suggestion
            # Get records from the raw response
```

File: tap_pendo/streams.py [hunk 7]
```diff
@@ -417,33 +425,44 @@ def lookback_window(self):
...
+            for item in ijson.items(resp.raw, 'results.item'):
+                yield humps.decamelize(item)
+            resp.close()
+        except (ConnectionError, ProtocolError, ReadTimeoutError, requests.exceptions.RequestException, Server42xRateLimitError) as e:
+            # Catch requestException and raise errors if we have to give up for certain conditions
+            if isinstance(e, requests.exceptions.RequestException) and to_giveup(e):
+                raise e from None
+            # Raise error if we have retried for 5 times
+            if count == 5:
+                LOGGER.info("Giving up request(...) after 5 tries (%s: %s)", e.__class__.__name__, str(e))
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tap_pendo/streams.py]: Updated the code with `LOGGER.error`

File: tap_pendo/streams.py (hunk 6 above)
Comment (harshpatel4crest) [reply tap_pendo/streams.py]: Updated the comment.

File: tap_pendo/streams.py (hunk 5 above)
Comment (harshpatel4crest) [reply tap_pendo/streams.py]: Updated the variable name to BACKOFF_FACTOR

File: tap_pendo/streams.py (hunk 1 above)
Comment (harshpatel4crest) [reply tap_pendo/streams.py]: Removed unused variable.

File: tap_pendo/streams.py (hunk 3 above)
Comment (harshpatel4crest) [reply tap_pendo/streams.py]: Updated the comment.

File: tap_pendo/streams.py (hunk 2 above)
Comment (harshpatel4crest) [reply tap_pendo/streams.py]: Updated the variable name.

File: tap_pendo/streams.py (hunk 1 above)
Comment (harshpatel4crest) [reply tap_pendo/streams.py]: Removed unused variable.

**Review by dbshah1212 — APPROVED** [review]

**Review by savan-chovatiya — APPROVED** [review]

File: tests/tap_tester/test_all_fields.py [hunk 8]
```diff
@@ -78,10 +80,10 @@ def test_run(self):
...
                 elif stream == "guide_events":
                     expected_all_keys = expected_all_keys - {'poll_response', "poll_id"}
-                elif stream == "features":
-                    expected_all_keys = expected_all_keys - {'page_id'}
-                elif stream == "guides":
-                    expected_all_keys = expected_all_keys - {'audience'}
+                # elif stream == "features":
+                #     expected_all_keys = expected_all_keys - {'page_id'}
+                # elif stream == "guides":
+                #     expected_all_keys = expected_all_keys - {'audience'}
```
Comment (RushiT0122) [inline tests/tap_tester/test_all_fields.py:86]: New records added in test account includes these fields to comment .

**Comment by Vi6hal** [comment]
Closing This PR, to clean up the old PR's.
as discussed with @RushiT0122 these changes were merged via a different PR.

also backoff for ConnectionResetError/ConnectionError exists in master branch
https://github.com/singer-io/tap-pendo/blob/master/tap_pendo/streams.py#L199-L202

**Pull request closed without merging**